test/boost/aws_error_injection_test.cc

```diff
@@ -0,0 +1,280 @@
+};
+
+static std::function<uint16_t(void)> get_port = [] { return 2012; };
+static std::function<std::string(void)> get_address = [] { return "127.0.0.1"; };
+
+struct unshare_fixture {
+    unshare_fixture() {
+        if (unshare(CLONE_NEWNET) != -1) {
+            std::system("ip l s lo up");
+            BOOST_TEST_MESSAGE("The network namespace was unshared");
+        } else {
+            BOOST_TEST_MESSAGE("Cant unshare network namespace. Randomizing addresses and ports");
+            get_port = []() {
+                std::random_device rd;
+                std::mt19937 gen(rd());
+                std::uniform_int_distribution<uint16_t> ports(std::numeric_limits<int16_t>::max(), std::numeric_limits<uint16_t>::max());
+                return ports(gen);
+            };
+            get_address = []() {
+                std::random_device rd;
+                std::mt19937 gen(rd());
+                std::uniform_int_distribution<uint8_t> octet;
+                return fmt::format("127.{}.{}.{}", octet(gen), octet(gen), octet(gen));
+            };
+        }
+    }
+};
+
+[[maybe_unused]] static unshare_fixture network_unshare;
+
+static s3::endpoint_config_ptr make_minio_config(uint16_t port) {
+    s3::endpoint_config cfg = {
+        .port = port,
+        .use_https = false,
+        .aws = {{
+            .access_key_id = "foo",
+            .secret_access_key = "bar",
+            .session_token = "baz",
+            .region = "us-east-1",
+        }},
+    };
+    return make_lw_shared<s3::endpoint_config>(std::move(cfg));
+}
+
+void test_client_upload_file(std::string_view test_name, const std::string& address, uint16_t port, size_t total_size, size_t memory_size) {
+    tmpdir tmp;
+    const auto file_path = tmp.path() / fmt::format("test-{}", ::getpid());
+
+    {
+        file f = open_file_dma(file_path.native(), open_flags::create | open_flags::wo).get();
+        auto output = make_file_output_stream(std::move(f)).get();
+        auto close_file = deferred_close(output);
+        std::string_view data = "1234567890ABCDEF";
+        // so we can test !with_remainder case properly with multiple writes
+        SCYLLA_ASSERT(total_size % data.size() == 0);
+
+        for (size_t bytes_written = 0; bytes_written < total_size; bytes_written += data.size()) {
+            output.write(data.data(), data.size()).get();
+        }
+    }
+
+    const auto object_name = fmt::format("/{}/{}-{}", "test", test_name, ::getpid());
+
+    semaphore mem{memory_size};
+    auto client = s3::client::make(address, make_minio_config(port), mem);
+    auto client_shutdown = deferred_close(*client);
+    client->upload_file(file_path, object_name).get();
+}
+
+SEASTAR_THREAD_TEST_CASE(test_multipart_upload_file_success) {
+    auto port = get_port();
+    auto address = get_address();
+    server server(server::failure_policy::SUCCESS);
+    server.start(address, port).get();
+    auto close_server = deferred_stop(server);
```

Comment:
should move this line up right after `server server(server::failure_policy::SUCCESS);`, as `server` (BTW, this does not compile with GCC, if we name a variable with the same name as a type) has a member variable named `http_server`, which creates a `distributed<http_server>` in its ctor instead of in its `start()` member function. so we have to define `deferred_stop(server)` after `server` is initialized. otherwise, if `start()` fails, we have the failure like:
```
Starting server on 127.0.0.1:12012
aws_error_injection_test: /home/kefu/dev/scylladb/seastar/include/seastar/core/sharded.hh:565: seastar::sharded<seastar::httpd::http_server>::~sharded() [Service = seastar::httpd::http_server]: Assertion `_instances.empty()' failed.
```